src/storage-controller/src/lib.rs

```diff
@@ -343,6 +344,18 @@ where
         Ok(result)
     }
 
+    fn clusters_caught_up(
+        &self,
+        _allowed_lag: Self::Timestamp,
+        _cutoff: Self::Timestamp,
+        _now: Self::Timestamp,
+        _live_frontiers: &BTreeMap<GlobalId, Antichain<Self::Timestamp>>,
+        _exclude_collections: &BTreeSet<GlobalId>,
+    ) -> bool {
+        // TODO(benesch): implement this.
+        true
+    }
```

Comment:
I wonder if the controllers are the right place to implement this logic. AIUI this logic (a) is specific to 0dt upgrades, (b) now needs to be duplicated in both controllers, and (c) only uses information also available to the coordinator (frontiers and number of replicas). It seems like it fits better somewhere in the coordinator, close to the code that does the promotion checking.

Reply:
You make a good point! It does seem to neatly fit into the coordinator. I put up https://github.com/MaterializeInc/materialize/pull/29714 with that refactoring.

Reply:
Indeed, this makes things much cleaner. This PR now has what would appear to be a working implementation of the caught up check for any sources that are running in read-only mode.